compio-dispatcher/src/lib.rs

```diff
@@ -86,30 +87,73 @@ impl Dispatcher {
         DispatcherBuilder::default()
     }
 
-    /// Dispatch a task to the threads.
+    fn prepare<Fut, Fn, R>(&self, f: Fn) -> (Executing<R>, Box<Closure>)
+    where
+        Fn: (FnOnce() -> Fut) + Send + 'static,
+        Fut: Future<Output = R> + 'static,
+        R: Any + Send + 'static,
+    {
+        let event = Event::new();
+        let handle = event.handle();
+        let res = Arc::new(Mutex::new(None));
+        let dispatched = Executing {
+            event,
+            result: res.clone(),
+        };
+        let closure = Box::new(|| {
+            Box::pin(async move {
+                *res.lock().unwrap() = Some(f().await);
+                handle.notify();
+            }) as BoxFuture<()>
+        });
+        (dispatched, closure)
+    }
+
+    /// Spawn a boxed closure to the threads.
+    ///
+    /// If all threads have panicked, this method will return an error with the
+    /// sent closure.
+    pub fn spawn(&self, closure: Box<Closure>) -> Result<(), SendError<Box<Closure>>> {
+        self.sender.send(closure)
+    }
+
+    /// Dispatch a task to the threads
     ///
     /// The provided `f` should be [`Send`] because it will be send to another
     /// thread before calling. The return [`Future`] need not to be [`Send`]
     /// because it will be executed on only one thread.
-    pub fn dispatch<
-        F: Future<Output = ()> + 'static,
-        Fn: (FnOnce() -> F) + Send + UnwindSafe + 'static,
-    >(
-        &self,
-        f: Fn,
-    ) -> io::Result<DispatcherJoinHandle> {
-        let event = Event::new();
-        let handle = event.handle();
-        let join_handle = DispatcherJoinHandle::new(event);
-        let closure = DispatcherClosure {
-            handle,
-            result: join_handle.result.clone(),
-            func: Box::new(|| f().boxed_local()),
-        };
-        self.sender
-            .send(closure)
-            .expect("the channel should not be disconnected");
-        Ok(join_handle)
+    ///
+    /// # Error
+    ///
+    /// If all threads have panicked, this method will return an error with the
+    /// sent closure.
+    pub fn dispatch<Fut, Fn>(&self, f: Fn) -> Result<(), SendError<Box<Closure>>>
```

Comment:
Is it possible to return `Result<(), SendError<Fn>>` here?

Reply:
No, we have made a wrapping layer for the `Fn` when dispatching.

Reply:
Then it will not be able to retrieve the fn again for a next dispatch operation.

Reply:
That's true, but since all threads have panicked it doesn't matter if we retry or not.